### DISCUSSION TITLE
Router::route_layer: Custom Basic authorization example

### DISCUSSION THREAD

**Discussion by phyber** [tokio-rs/axum#1742] [Q&A] [answered]
- [x] I have looked for existing issues (including closed) about this

## Feature Request

Hi,

I'm wondering if a `RequireAuthorizationLayer::custom` authorization example using the `Basic` scheme could be added.

Apologies if requests for documentation enhancement aren't considered features, and thanks for your time.

### Motivation

I've been trying to implement a custom `Basic` authorization layer in `axum` for a while, but the current documentation and examples haven't helped, and I've failed to implement it so far.

The use case I have that doesn't let me use the existing `RequireAuthorizationLayer::basic` is that I have a list of users to authenticate against, not just a single user.

I am coming from `actix_web` and `actix_web_httpauth` where the equivalent functionality was very simple to setup by comparison.

### Proposal

An example of using the `RequireAuthorizationLayer::custom` feature should be added.

### Alternatives

  - There is an `axum-auth` module, however, this module does not appear to offer the functionality I'm after.
  - There is a `jwt` example, extracting `Bearer` tokens and using those, somehow. It seems a bit magical, isn't using `layer_route`, and is completely undocumented, making it hard to know what's going on.
    - Attempting to implement similar extractors on `Basic` instead of `Bearer`, with the same patterns, results in errors that this example doesn't have, which is frustrating. Perhaps the example is using older versions of axum or its dependencies
  - I've attempted to search GitHub to see if anyone has implemented something similar, but either code search is bad, or I'm just not giving it the right prompts
  - I could just stick with `actix_web` which works, but Axum has resulted in a simpler implementation for much of the rest of the application, only the Basic authentication is difficult

**Comment by davidpdrsn** [comment, marked as answer]
I don't think `RequireAuthorizationLayer::custom` offers much when you're using axum. They're overdue for a rework I'd say.

I recommend you handle it yourself with `axum::middleware::from_fn`.

Also note that `axum-auth` is an external library and not maintained by the axum team.

**Reply by phyber** [reply]
Thanks! `middleware::from_fn` and its `_with_state` companion look like exactly what I was after.
